python/lsst/afw/multiband.py

```diff
@@ -0,0 +1,346 @@
+                filterNames = [filterIndex]
+            filterIndices = [self.filters.index(f) for f in filterNames]
+        return tuple(filterNames), filterIndices
+
+    def imageIndicesToNumpy(self, sliceArgs):
+        """Convert slicing format to numpy
+
+        LSST `afw` image-like objects use an `[x,y]` coordinate
+        convention, accept `Point2I` and `Box2I`
+        objects for slicing, and slice relative to the
+        bounding box `XY0` location;
+        while python and numpy use the convention `[y,x]`
+        with no `XY0`, so this method converts the `afw`
+        indices or slices into numpy indices or slices
+
+        Parameters
+        ----------
+        sliceArgs: `sequence`, `Point2I` or `Box2I`
+            An `(xIndex, yIndex)` pair, or a single `(xIndex,)` tuple,
+            where `xIndex` and `yIndex` can be a `slice` or `int`,
+            or list of `int` objects, and if only a single `xIndex` is
+            given, a `Point2I` or `Box2I`.
+
+        Returns
+        -------
+        y: index or slice
+            Index or slice in the y dimension
+        x: index or slice
+            Index or slice in the x dimension
+        bbox: Box2I
+            Bounding box of the image.
+            If `bbox` is `None` then the result is a point and
+            not a subset of an image.
+        """
+        # Use a common slicing algorithm as single band images
+        x, y, origin = interpretSliceArgs(sliceArgs, self.getBBox)
+
+        if origin == PARENT:
+            if isinstance(x, slice):
+                assert isinstance(y, slice)
+                bbox = Box2I(Point2I(x.start, y.start), Extent2I(x.stop-x.start, y.stop-y.start))
+                x = slice(x.start - self.x0, x.stop - self.x0)
+                y = slice(y.start - self.y0, y.stop - self.y0)
+            else:
+                x = x - self.x0
+                y = y - self.y0
+                bbox = None
+            return y, x, bbox
+        elif origin != LOCAL:
+            raise ValueError("Unrecognized value for origin")
+
+        # Use a local bounding box
+        if isinstance(x, slice):
+            assert isinstance(y, slice)
+            bbox = Box2I(Point2I(x.start + self.x0, y.start + self.y0),
+                         Extent2I(x.stop-x.start, y.stop-y.start))
+        else:
+            bbox = None
+        return y, x, bbox
+
+    def shiftedTo(self, xy0):
+        """Shift the bounding box but keep the same Extent
+
+        Parameters
+        ----------
+        xy0: `Point2I`
+            New minimum bounds of the bounding box
+        """
+        self._bbox = Box2I(xy0, self._bbox.getDimensions())
+        for singleObj in self.singles:
+            singleObj.setXY0(xy0)
+
+    def shiftedBy(self, offset):
+        """Shift a bounding box by an offset, but keep the same Extent
+
+        Parameters
+        ----------
+        offset: `Extent2I`
+            Amount to shift the bounding box in x and y.
+        """
```

Comment:
Ah, I didn't realize that's what it was supposed to do, although looking back at the RFC now the name makes a lot more sense... I'll fix this for all of the classes.

Reply:
Oh, I may have mispoken a bit between this comment and the one above.  On the single-band objects, `setXY0` *does* operate in-place (as should any method with that name).  The names `shiftedBy` and `shiftedTo` (by being past tense) imply that they return new objects.

Reply:
Arg, this is a breaking change until RFC-343 is implemented. Specifically, since the single band objects don't have `shiftedBy` and `shiftedTo` methods, the generic methods in `MultibandBase` don't know how to make a shallow copy of the `singles` and then modify only their bounding box while keeping a view to the original object.

So I'll leave the `shiftedBy` and `shiftedTo` methods in place (but with a `NotImplementedError` raised before the code executes, and also re-implement `setXy0` with a note in DM-10781 to also fix the multiband classes.